### PR TITLE
tests: add env file config coverage

### PR DESCRIPTION
**Pull request by Kludex**
Lines covered:
```python
        if env_file is not None:
            from dotenv import load_dotenv

            logger.info("Loading environment from '%s'", env_file)
            load_dotenv(dotenv_path=env_file)
```
File: `config.py`

### CODE REVIEW

**Comment by Kludex** [comment]
ok... The very informative comment on https://github.com/encode/uvicorn/pull/470 makes sense now hahaha

**Comment by Kludex** [comment]
I've found this: https://www.tutorialfor.com/questions-105322.htm

But I'm not sure how to act here... Any thoughts? @euri10

**Comment by euri10** [comment]
I dont remember I had issue on this, you can take a look at https://github.com/encode/uvicorn/pull/626/files which did part of coverage increase last year and was reverted.

File: tests/test_config.py [hunk 1]
```diff
@@ -247,3 +249,37 @@ def test_log_config_file(mocked_logging_config_module):
...
+    yield request.param
+    if os.getenv("FORWARDED_ALLOW_IPS"):
+        del os.environ["FORWARDED_ALLOW_IPS"]
+
+
+def test_env_file(web_concurrency: int, forwarded_allow_ips: str, caplog):
+    """
+    Test that one can load environment variables using an env file.
+    """
+    with tempfile.NamedTemporaryFile("w+") as fp:
```
Comment (euri10) [inline tests/test_config.py]: why not use pytest fixtures for temporary files ?

File: tests/test_config.py [hunk 2]
```diff
@@ -247,3 +249,37 @@ def test_log_config_file(mocked_logging_config_module):
...
+        content = (
+            f"WEB_CONCURRENCY={web_concurrency}\n"
+            f"FORWARDED_ALLOW_IPS={forwarded_allow_ips}\n"
+        )
+        fp.write(content)
+        fp.seek(0)
+        with caplog.at_level(logging.INFO):
```
Comment (euri10) [inline tests/test_config.py]: why not following other tests pattern with config.load ?

Comment (Kludex) [reply tests/test_config.py]: The test that checks if should reload (above on this file), is not calling it as well. So I thought it was fine?

But I'll add it, sure. 😗

File: tests/test_config.py (hunk 1 above)
Comment (Kludex) [reply tests/test_config.py]: Thanks, going to apply later.

File: tests/test_config.py (hunk 2 above)
Comment (Kludex) [reply tests/test_config.py]: Applied.

File: tests/test_config.py (hunk 1 above)
Comment (Kludex) [reply tests/test_config.py]: Applied.

**Comment by Kludex** [comment]
> I dont remember I had issue on this, you can take a look at https://github.com/encode/uvicorn/pull/626/files which did part of coverage increase last year and was reverted.

Oops... I've read this after applying the changes. Is my implementation enough or should I match the previous PR?

**Comment by euri10** [comment]
I'm not sure we need to test for a permutation of both `WEB_CONCURRENCY` and `FORWARDED_ALLOW_IPS` but would there be a good reason we need more explicit test names with `ids` or similar, the below is way too abstruse

20210314_1632_232x133_1615735975

so either keep it this way but name the tests more explicitely or just run 2 tests, one for the 1st env var, one for the 2nd one, hope this makes sense.

**Comment by Kludex** [comment]
Do you mind sharing the way to see that cool interface (on the screenshot), please? :pleading_face:

**Comment by euri10** [comment]
> Do you mind sharing the way to see that cool interface (on the screenshot), please?

it's just Pycharm ! on a side note using `pytest -rA` should show you the same

File: tests/test_config.py [hunk 3]
```diff
@@ -249,6 +250,40 @@ def test_log_config_file(mocked_logging_config_module):
...
+
+
+@pytest.fixture(params=["127.0.0.1", "127.0.0.2"])
+def forwarded_allow_ips(request):
+    yield request.param
+    if os.getenv("FORWARDED_ALLOW_IPS"):
+        del os.environ["FORWARDED_ALLOW_IPS"]
+
+
+def test_env_file(web_concurrency: int, forwarded_allow_ips: str, caplog, tmpdir):
```
Comment (euri10) [inline tests/test_config.py]: I would use a `tmp_path` fixture instead, I'm not against the tmpdir usage though it's just that I feel it's cleaner to use Path-like objects.
I'll approve and let you choose !

**Review by euri10 — APPROVED** [review]

File: tests/test_config.py (hunk 3 above)
Comment (Kludex) [reply tests/test_config.py]: thanks

**Pull request merged**